web-js/src/hooks/useTodos_graphql.js

```diff
@@ -0,0 +1,197 @@
+import React from "react";
+import { ApolloClient, gql, HttpLink, InMemoryCache } from "@apollo/client";
+import { useWatch } from "./useWatch";
+import { useCollection } from "./useCollection";
+import { useRealmApp } from "../components/RealmApp";
+import {
+  addValueAtIndex,
+  replaceValueAtIndex,
+  updateValueAtIndex,
+  removeValueAtIndex,
+  getTodoIndex,
+} from "../utils";
+
+function useApolloClient() {
+  const realmApp = useRealmApp();
+  if (!realmApp.currentUser) {
+    throw new Error(`You must be logged in to Realm to call useApolloClient()`);
+  }
+
+  const client = React.useMemo(() => {
+    const graphqlUri = `https://realm.mongodb.com/api/client/v2.0/app/${realmApp.id}/graphql`;
+    // Local apps should use a local URI!
+    // const graphqlUri = `https://us-east-1.aws.stitch.mongodb.com/api/client/v2.0/app/${realmApp.id}/graphql`
+    async function getValidAccessToken() {
+      // An already logged in user's access token might be stale. To guarantee that the token is
+      // valid, we refresh the user's custom data which also refreshes their access token.
+      await realmApp.currentUser.refreshCustomData();
+      return realmApp.currentUser.accessToken;
+    }
+    return new ApolloClient({
+      link: new HttpLink({
+        uri: graphqlUri,
+        // We define a custom fetch handler for the Apollo client that lets us authenticate GraphQL requests.
+        // The function intercepts every Apollo HTTP request and adds an Authorization header with a valid
+        // access token before sending the request.
+        fetch: async (uri, options) => {
+          const accessToken = await getValidAccessToken();
+          options.headers.Authorization = `Bearer ${accessToken}`;
+          return fetch(uri, options);
+        },
+      }),
+      cache: new InMemoryCache(),
+    });
+  }, [realmApp.currentUser, realmApp.id]);
+
+  return client;
+}
+
+export function useTodos() {
+  // Get a graphql client and set up a list of todos in state
+  const realmApp = useRealmApp();
+  const graphql = useApolloClient();
+  const [todos, setTodos] = React.useState([]);
+  const [loading, setLoading] = React.useState(true);
+
+  // Fetch all todos on load and whenever our graphql client changes (i.e. either the current user OR App ID changes)
+  React.useEffect(() => {
+    const query = gql`
+      query FetchAllTodos {
+        tasks {
+          _id
+          _partition
+          isComplete
+          summary
+        }
+      }
+    `;
+    graphql.query({ query }).then(({ data }) => {
+      setTodos(data.tasks);
+      setLoading(false);
+    });
+  }, [graphql]);
+
+  // Use a MongoDB change stream to reactively update state when operations succeed
+  const taskCollection = useCollection({
+    service: "mongodb-atlas",
+    db: "todo",
+    collection: "Task",
+  });
+  useWatch(taskCollection, {
+    onInsert: (change) => {
```

Comment:
Do we offer `subscriptions` in our graphql API?  Seems like the GraphQL implementation here is a hybrid of MQL and GraphQL given this collection watcher.